### DISCUSSION TITLE
Supporting collections of free functions  using modules

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#5310] [Language Ideas] [answered]
Point of clarification: "modules" here are unrelated to [netmodules](https://docs.microsoft.com/en-us/archive/blogs/junfeng/netmodule-vs-assembly).

I often find myself writing collections of functions in many projects. These functions have a common shape: they tend to be deterministic, they tend to be used throughout the code and they are referenced directly by utilising `using static` in classes that reference them.

Because C# is a type-based language, these functions exist as static methods within static classes. This creates two issues:
1. The minor one in that `static` appears numerous times throughout the declaration.
2. The more significant one in that gathering a collection of functions together into a static class feels semantically wrong. They aren't used as class methods, they are used as "free functions", especially since that `static using` allows the class to be partially hidden from the developer.

With C# 10, we get global usings. This means I can now add a `global using static SomeFunctions` to the file containing the definition of `SomeFunctions`. So that class is hidden away even more. But the class still exists and it still has that semantic incongruity about it. Further, the notion of it being a set of free functions is easily broken by eg making the class non-static, adding properties, sub classes to it, etc.

So I'd like to propose that C# adopt the concept of a "module", which is just a collection of functions, which might look something like:
```c#
namespace X;

internal module FooBar
{
    int Foo1X(int x) => Bar1(x, 0);
    int Foo1Y(int y) => Bar1(0, y);

    int Foo2X(int x) => Bar2(x, 0);
    int Foo2Y(int y) => Bar2(0, y);

    private int Bar1(int x, int y) => Bar(x, y, true);
    private int Bar2(int x, int y) => Bar(x, y, false);
    
    private int Bar(int x, int y, bool flag) => 0; // or whatever
}
```
A module would not be able to contain types, properties, non-private fields (or maybe no fields at all), events, delegates etc. And eg `Foo1X` would be a "free function" that can be referenced directly without having to express it as `FooBar.Foo1X`.

In terms of implementation, the above would be lowered to:
```c#
global using static X.FooBar;

namespace X;

internal static class FooBar
{
    internal static int Foo1X(int x) => Bar1(x, 0);
    internal static int Foo1Y(int y) => Bar1(0, y);
    internal static int Foo1(int x, int y) => Bar1(x, y);

    internal static int Foo2X(int x) => Bar2(x, 0);
    internal static int Foo2Y(int y) => Bar2(0, y);
    internal static int Foo2(int x, int y) => Bar2(x, y);

    private static int Bar1(int x, int y) => Bar(x, y, true);
    private static int Bar2(int x, int y) => Bar(x, y, false);
    
    private static int Bar(int x, int y, bool flag) => 0;
}
```
So no runtime changes are needed, it simply becomes semanic sugar hiding away the class as an implementation detail.

**Comment by YairHalberstadt** [comment]
What's the purpose of modules? Why not support top level functions directly

**Reply by Richiban** [reply]
You *could* have bare functions within a namespace, but a module allows you to have private methods that are only visible within that module.

**Reply by DavidArno** [reply]
@Richiban beat me to it. but to reinforce his point, the reason for having the module is to constrain the scope of private functions. I guess they could be effectively scoped to the file (or compilation unit) if the rules around only one compilation unit being allowed to contain top-level statements were relaxed though.

**Comment by Richiban** [comment]
I'm not sure I like the automatic `global using`; it's incongruous with the rest of C# and it also kind of defies the point of giving the module a name at all. Why not have the module behave just like a namespace? You can always add the `global using` yourself and then you open the module, you get the functions.

To be honest, while I think C# does deserve the concept of a module I think it's only really a minor syntactic improvement over a static class (and some restrictions).

**Reply by DavidArno** [reply]
> I think it's only really a minor syntactic improvement...

This is true. But to my mind it is a significant *semantic* improvement as it makes it clear that these are effectively "free functions" and the container is unimportant, possibly to the extent of not even requiring a name for the module. The syntax improvements are really secondary to that.

**Reply by Richiban** [reply]
If we run with this idea for the moment, I'd like to propose merging the idea of a module and a namespace, so your original code would look like this:

```cs
internal module X.FooBar
{
    int Foo1X(int x) => Bar1(x, 0);
    int Foo1Y(int y) => Bar1(0, y);

    int Foo2X(int x) => Bar2(x, 0);
    int Foo2Y(int y) => Bar2(0, y);

    private int Bar1(int x, int y) => Bar(x, y, true);
    private int Bar2(int x, int y) => Bar(x, y, false);
    
    private int Bar(int x, int y, bool flag) => 0; // or whatever
}
```

or, if we also allow a mirror of the file-scoped namespaces:

```cs
module X.FooBar;

int Foo1X(int x) => Bar1(x, 0);
int Foo1Y(int y) => Bar1(0, y);

int Foo2X(int x) => Bar2(x, 0);
int Foo2Y(int y) => Bar2(0, y);

private int Bar1(int x, int y) => Bar(x, y, true);
private int Bar2(int x, int y) => Bar(x, y, false);

private int Bar(int x, int y, bool flag) => 0; // or whatever
```

**Comment by HaloFour** [comment]
Could mimic this with source generators.  Stick an attribute on a static class and have a separate file created with the `global using static` statements that make the members accessible directly through the project.

**Reply by DavidArno** [reply]
How would that address the main semantic driver of wanting to remove the need to declare free functions within a class? The `global using static` feature just makes the implementation easier.

So I disagree: I could use use a source generator to auto-create the `global using static`, but that's unrelated to this proposal.

**Reply by HaloFour** [reply]
I don't see much difference in the keyword `module` vs. `static class`, personally.  If the argument was to have the methods defined truly at the top level and completely outside of some scope I could see that more, but I think the benefit would be more in the consumption of the functions, not their declarations, given the ratio of the frequency between the two.

**Reply by jnm2** [reply]
Indeed, VB's keyword for `static class` is `Module`.

**Reply by DavidArno** [reply]
@jnm2, VB implements its Module functionality via static classes in the runtime.

You are focusing on the implementation, not the feature in the way you express it.

**Reply by HaloFour** [reply]
IMO the implementation isn't all that far removed from the syntax you'd write anyway.  Is there really that much benefit to add a new keyword `module` compared to how you could achieve the same exact functionality in C# 10?

```cs
global using static X.FooBar;

namespace X;

internal static class FooBar
{
    static int Foo1X(int x) => Bar1(x, 0);
    static int Foo1Y(int y) => Bar1(0, y);

    static int Foo2X(int x) => Bar2(x, 0);
    static int Foo2Y(int y) => Bar2(0, y);

    private static int Bar1(int x, int y) => Bar(x, y, true);
    private static int Bar2(int x, int y) => Bar(x, y, false);
    
    private static int Bar(int x, int y, bool flag) => 0; // or whatever
}
```

**Reply by jnm2** [reply]
@DavidArno My point is that I think we _have_ modules. I'm not sure why seeing `static` written fewer times is something that warrants a new feature, and the additional restrictions don't seem compelling to me either yet.

**Reply by DavidArno** [reply]
@jnm2, as I tried to explain in the OP, seeing static written a few less times is just a side benefit. The aim of the proposal was to make the intent of the code clearer.

A comparable feature is read-only fields. C# already had read-only fields: the developer didn't write to them. They were read-only by implication. But the `readonly` keyword enforces the intent and makes it clear to other developers that it's supposed to be read-only. Likewise, a module would be clear indication that this is a supposed to be collection of functions rather than a static class that currently only has methods.

**Reply by HaloFour** [reply]
> They were read-only by implication.

Except when the accessibility level is anything except `private` in which case the metadata is required at the IL level to indicate that the field cannot be written to.  I think this is a poor analogy.

> a module would be clear indication that this is a supposed to be collection of functions rather than a static class that currently only has methods.

The bigger question is _why_ there needs to be any such indication, or why those imposed limitations are necessary?  Many other languages with modules permit top-level state, like F# or Kotlin.

**Comment by jnm2** [comment]
> A module would not be able to contain types, properties, non-private fields (or maybe no fields at all), events, delegates etc.

What happens on the day that someone asks for one of these? Why are these restrictions good?

**Reply by DavidArno** [reply]
The day someone wants to add a property to a module, they no longer have a collection of free functions, they have a static class. So they'd change it to a static class to make that clear. The restrictions are deliberate as they focus the feature on just being a way of creating a  collection of free functions, rather than it just being syntactic sugar that reduces the occurrence of `static`.

**Reply by jnm2** [reply]
What useful thing is achieved by seeing at a glance that a type is a module and thus does not have state? (And this would only be visible in IDEs, since the call sites all look the same as if you were calling into a static class.)

Adding a field or property to a static class does not itself tell you anything about the functions in the static class. Functions in the proposed `module` could read static fields declared in non-module types already.

**Comment by huoyaoyuan** [comment]
Generally I don't think auto importing is a good practice. I'm not a fan of `global using` at all.

**Comment by masonwheeler** [comment]
This is what `System.Math` should have been since Day 1.  There's a certain cargo-cultishness to the design of Java, which then infected C# 1.0 and we've been stuck with it ever since, that everything must be shoehorned into a class whether it actually needs one or not.  This proposal would go a long way towards remedying that decades-old mistake.

**Reply by HaloFour** [reply]
Considering that this proposal seeks to only have `module Math` be shorthand for `static class Math` I fail to see what significance doing so would have to the language or ecosystem.  You still have to put the functions into a named _something_.  Whether it's a class or module seems to have no practical difference whatsoever.

**Reply by DavidArno** [reply]
This proposal most certainly doesn't "*seek to only have module Math be shorthand for static class Math*". The fact you claim this shows that you haven't understood the proposal at all. Please try to take the time to think through an idea before you rubbish it via a crude strawman.

**Reply by HaloFour** [reply]
That is the effective difference, unless you want to amend your implementation details.  The concern seems to be much more about what it's called vs. what it does.

**Reply by jnm2** [reply]
@masonwheeler Can you explain why static methods in classes are not the obvious existing place to put free functions? This proposal would have call sites look exactly the same as though they were in classes.

**Reply by masonwheeler** [reply]
@jnm2 A class is a blueprint for an object, which is a collection of shared state bound together with functions that operate on that state. If there is no shared state, and there is no object at all (because it's a "static class") then why should it even be a class in the first place?

Anytime you encounter a static class, you're seeing a leak in the abstraction poking out.

**Reply by jnm2** [reply]
I guess that's what I'm asking. If the only thing you'd like to have changed is that the syntax `static class` would have been `module`, the way VB does it, then I'd say we *have* modules today.

**Reply by jnm2** [reply]
Also, I do like the principle that if it's okay for a static method to exist in a non-static class with instance members, it's just as okay for it to continue existing in a non-static class with no instance members. And from there it's only a very small step to me to add `static` to the class as a courtesy to say "it's not interesting to instantiate or inherit from this."

**Reply by masonwheeler** [reply]
Are you aware of any non-static classes with no instance members?  Other than arguably `System.Object` (which does have instance state; it just hides it under CLR voodoo) I can't think of any off the top of my head.

**Reply by jnm2** [reply]
Yes, but I have refactored most I've come across to `static class` as a courtesy, since that only has benefits. The progression I mentioned is still one that feels logical and positive to me regardless of whether a file was captured in that intervening state.

**Reply by HaloFour** [reply]
We're well past the point of litigating whether or not to add the `static` modifier to classes.   The question at this point is what functionally would differ a `module` from a `static class` other than the name.

**Reply by CyrusNajmabadi** [reply]
Agreed. Is this just a syntactic shorthand?  Of so, is rather use 'module' for something substantively different.

**Comment by RikkiGibson** [comment, marked as answer]
I feel like it might be OK to remove the requirement to include a 'static' modifier on static class members, and simply make the members implicitly static.

**Reply by Richiban** [reply]
I'd love that! I never really understood the requirement in the first place; having to declare each member as `static` was redundant and just added to the noise in the file.

**Reply by DavidArno** [reply]
@Richiban,

The common argument is that if you remove the `static` from in front of `class`, then all those methods become instance methods. It's not an argument that has ever made sense to me, but I think the language team find it compelling. So that likely will never happen.

**Reply by HaloFour** [reply]
> I feel like it might be OK to remove the requirement to include a 'static' modifier on static class members, and simply make the members implicitly static.

Other than saving a tiny bit of typing (which autocomplete will largely do for you) is there really much benefit for doing that?  IMO it would make it harder to parse the nature of the members when reading a class definition, I'd always have to scan back to see whether or not the class was `static`.

> I'd love that! I never really understood the requirement in the first place

It was a guardrail to help in avoiding bugs where you intended the class to only contain static members and not be instantiatable/inheritable.

**Reply by DavidArno** [reply]
I find unit tests a much better "guardrail" than noise in my source. But each to their own, I guess 😆

**Reply by masonwheeler** [reply]
> The common argument is that if you remove the `static` from in front of `class`, then all those methods become instance methods. It's not an argument that has ever made sense to me, but I think the language team find it compelling. So that likely will never happen.

I say, *why is that a bad thing?*  If you're turning a static class into a real class, it's happening for a reason.

I actually did this earlier this week.  I a static class that I decided made more sense as a real class, in part because it was passing around enough state between its methods that it would be simpler just to turn those into fields.  And as part of the process, I had to go through and *remove* the explicit `static` tag from all these methods just to get it to work.  It would have been a lot nicer if they had never needed to be there in the first place.

**Reply by HaloFour** [reply]
@DavidArno 

I much rather a compiler find and catch simple bugs like that.  If I wanted to delegate that to unit tests I'd write JavaScript.  😁

**Reply by DavidArno** [reply]
@HaloFour, touché! 😆

**Reply by jnm2** [reply]
@masonwheeler The nice part is that I can select `static ` , and then either use multicursor select matching + delete or use "replace all" and be finished in one or two seconds.

**Reply by RikkiGibson** [reply]
> @Richiban,
> 
> The common argument is that if you remove the `static` from in front of `class`, then all those methods become instance methods. It's not an argument that has ever made sense to me, but I think the language team find it compelling. So that likely will never happen.

It's not something I feel deeply passionate about, but it does feel a little odd that we ended up making an inconsistent decision when it comes to `readonly struct`. We do require readonly on the fields, but don't require readonly on other members--this is necessary for compat since the features didn't arrive at the same time, but in usage I felt there was some kind of line with adding readonly on the type and methods for example which made it just feel *too* redundant.

**Reply by HaloFour** [reply]
@RikkiGibson 

That doesn't make sense.  Marking a `struct` as `readonly` requires that the state of the struct not be modifiable.  This is accomplished by marking the fields as `readonly`, not by marking the methods as `readonly`.  Forcing the developer to also mark the methods as readonly wouldn't add anything.

**Reply by RikkiGibson** [reply]
Yeah, I guess what I'm saying is: in the alternate universe where￼ readonly structs and readonly members arrived at the same time, we still would probably not require you to put readonly on a method within a readonly struct.￼ It's too onerous. The method obviously has a readonly `this`.￼

The question this raises to me is simply: why, then, do we require the same form of repetition between a static class and a static method within it?

**Reply by CyrusNajmabadi** [reply]
Mainly, it was too be very obvious what you were looking at. However, that said, I imagine we're not 100% wedded to it, and it could be relaxed if a compelling case were made! :-)

**Reply by DavidArno** [reply]
@HaloFour 
> Forcing the developer to also mark the methods as readonly wouldn't add anything.

Yes it would: it would ensure that should the developer remove `readonly` from the struct, the method marked `readonly` cannot suddenly become a method that can mutate state. It is directly equivalent to the argument used to justify having to repeatedly specify `static` and is language inconsistency as @RikkiGibson. 

However, I'm not complaining about not needing to specify `readonly` everywhere as I agree with "*Forcing the developer to also mark the methods as readonly wouldn't add anything*". Which is why I agree that forcing the developer to also mark the methods as static doesn't add anything...

**Reply by DavidArno** [reply]
As I don't seem to be able to get a debate on the benefits of new syntax to better express the semantics of collections of free functions going, I think I need to accept that the debate that most people want is removing noise from static classes vs seeing the requirement to repeatedly specify static as a benefit. So I'm going to make it answered by this answer and move then. Thanks all.

**Reply by HaloFour** [reply]
> the method marked `readonly` cannot suddenly become a method that can mutate state

It couldn't anyway, not unless you also removed the `readonly` modifier from one or more of the fields.  The type doesn't become less readonly when you remove the guardrail, specifically because the `readonly` modifier on the type does not automatically extend to the fields.  This is the same as with `static`, where the modifier on the type indicates that instances of the type are impossible, which makes instance members illegal but does not change the manner in which the members are declared.

As for whether or not the requirements around the members of static types should be changed, I just don't find that particularly interesting.  Autocomplete writes most of that for you, and I think the verbosity is a small trade-off for clarity.  What would be more interesting out of this proposal is the consumption of said functions.  But with both `using static` and `global using static` my opinion is that we're already functionally there with "modules", they're just called "static classes", so the debate becomes whether there is value in a second syntax to declare what you can already declare.

**Reply by jnm2** [reply]
> It couldn't anyway, not unless you also removed the readonly modifier from one or more of the fields.

Can't you still assign to `this` even then?

**Reply by HaloFour** [reply]
@jnm2 

> Can't you still assign to `this` even then?

Crazily enough, yeah.  So the `readonly` modifier on the type does make one difference.  But so does that `static` modifier on the type, in that the type is implicitly made `sealed` with a private ctor.
